File: analytics/src/main/java/io/appmetrica/analytics/impl/DeeplinkConsumer.java
package io.appmetrica.analytics.impl;

import android.content.Intent;
import android.text.TextUtils;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.appmetrica.analytics.coreutils.internal.logger.YLogger;
import io.appmetrica.analytics.impl.utils.ApiProxyThread;
import java.util.HashMap;
import java.util.Map;

public class DeeplinkConsumer {

    private static final String TAG = "[DeeplinkConsumer]";

    @NonNull
    private final IMainReporter mainReporter;
    @NonNull
    private final Map<Boolean, String> deeplinksBySource = new HashMap<Boolean, String>();

    public DeeplinkConsumer(@NonNull IMainReporter mainReporter) {
        this.mainReporter = mainReporter;
    }

    @ApiProxyThread
    public void reportAppOpen(@Nullable final Intent intent) {
        if (intent != null) {
            String deeplink = intent.getDataString();
            reportAppOpenInternal(deeplink, false);
        }
    }

    @ApiProxyThread
    public void reportAutoAppOpen(@Nullable final String deeplink) {
        reportAppOpenInternal(deeplink, true);
    }

    @ApiProxyThread
    public void reportAppOpen(@Nullable final String deeplink) {
        reportAppOpenInternal(deeplink, false);
    }

    @ApiProxyThread
    private void reportAppOpenInternal(@Nullable String deeplink, boolean auto) {
        YLogger.info(TAG, "Try to report deeplink: %s, auto: %b", deeplink, auto);
        if (TextUtils.isEmpty(deeplink) == false) {
            if (Utils.areEqual(deeplink, deeplinksBySource.get(!auto)) == false) {
                mainReporter.reportAppOpen(deeplink, auto);
            } else {
                YLogger.info(TAG, "Deeplink %s has already been reported from another source. " +
                        "Skipping.", deeplink);
            }
            deeplinksBySource.put(auto, deeplink);
        }
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/DeeplinkConsumer.kt
package io.appmetrica.analytics.impl

import android.content.Intent
import io.appmetrica.analytics.impl.utils.ApiProxyThread
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger

class DeeplinkConsumer(private val mainReporter: IMainReporter) {

    private val tag = "[DeeplinkConsumer]"
    private var lastDeeplink = ""

    @ApiProxyThread
    fun reportAppOpen(intent: Intent?) {
        if (intent != null) {
            val deeplink = intent.dataString
            reportAppOpenInternal(deeplink, false)
        }
    }

    @ApiProxyThread
    fun reportAutoAppOpen(deeplink: String?) {
        reportAppOpenInternal(deeplink, true)
    }

    @ApiProxyThread
    fun reportAppOpen(deeplink: String?) {
        reportAppOpenInternal(deeplink, false)
    }

    @ApiProxyThread
    private fun reportAppOpenInternal(deeplink: String?, auto: Boolean) {
        DebugLogger.info(tag, "Try to report deeplink: $deeplink, auto: $auto")
        deeplink?.takeIf { it.isNotEmpty() } ?: return
        if (lastDeeplink != deeplink) {
            lastDeeplink = deeplink
            mainReporter.reportAppOpen(deeplink, auto)
        } else {
            DebugLogger.info(tag, "Ignore duplicated deeplink: $deeplink")
        }
    }
}
